File: demo/src/main/java/com/example/ConsumidorExpress.java
package com.example;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.DeliverCallback;

/**
 * Consumidor de Pedidos EXPRESS
 *
 * Se suscribe unicamente a mensajes con routing key "pedido.express".
 * Simula el procesamiento de pedidos con envio en 24 horas.
 */
public class ConsumidorExpress {

    private static final String EXCHANGE_NAME = "pedidos_exchange";
    // Cola exclusiva para pedidos express
    private static final String QUEUE_NAME = "cola_express";
    // Patron que acepta solo pedidos express
    private static final String ROUTING_KEY = "pedido.express";

    public static void main(String[] args) throws Exception {

        // 1. Conexion a RabbitMQ
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("localhost");
        factory.setPort(5672);

        Connection connection = factory.newConnection();
        Channel channel = connection.createChannel();

        // 2. Declarar el exchange (debe coincidir con el productor)
        channel.exchangeDeclare(EXCHANGE_NAME, "topic");

        // 3. Declarar la cola y vincularla al exchange con la routing key
        channel.queueDeclare(QUEUE_NAME, true, false, false, null);
        channel.queueBind(QUEUE_NAME, EXCHANGE_NAME, ROUTING_KEY);

        System.out.println("[ConsumidorExpress] Esperando pedidos express. CTRL+C para salir.");

        // 4. Callback: se ejecuta por cada mensaje recibido
        DeliverCallback deliverCallback = (consumerTag, delivery) -> {
            String mensaje = new String(delivery.getBody(), "UTF-8");
            System.out.println("\n[ConsumidorExpress] Pedido recibido: " + mensaje);
            // Simular el procesamiento del pedido express
            procesarPedidoExpress(mensaje);
        };

        // 5. Iniciar consumo (auto-ack = true para simplificar el ejemplo)
        channel.basicConsume(QUEUE_NAME, true, deliverCallback, consumerTag -> {
        });
    }

    /**
     * Simula el flujo de procesamiento de un pedido express:
     * verificacion de pago, asignacion de mensajero y confirmacion.
     */
    private static void procesarPedidoExpress(String mensaje) {
        System.out.println("  -> [EXPRESS] Verificando pago...");
        System.out.println("  -> [EXPRESS] Preparando paquete para envio en 24h...");
        System.out.println("  -> [EXPRESS] Asignando mensajero prioritario...");
        System.out.println("  -> [EXPRESS] Confirmacion enviada al cliente. HECHO.");
    }
}

File: demoPy/consumidor_express.py
import pika
import json

"""
Consumidor de Pedidos EXPRESS - Python
========================================
Se suscribe al exchange 'pedidos_exchange' con la routing key 'pedido.express'.
Solo recibe y procesa pedidos marcados como express (entrega en 24h).
"""

EXCHANGE_NAME = "pedidos_exchange"
QUEUE_NAME    = "cola_express"
ROUTING_KEY   = "pedido.express"


def procesar_pedido_express(pedido: dict):
    """
    Simula el flujo de procesamiento de un pedido express:
    verificacion de pago, preparacion y asignacion de mensajero.
    """
    print(f"  -> [EXPRESS] Verificando pago del pedido #{pedido['id']}...")
    print(f"  -> [EXPRESS] Preparando '{pedido['producto']}' (x{pedido['cantidad']}) para envio en 24h...")
    print(f"  -> [EXPRESS] Asignando mensajero prioritario...")
    print(f"  -> [EXPRESS] Confirmacion enviada al cliente. HECHO.\n")


def callback(ch, method, properties, body):
    """Funcion de callback: se ejecuta cada vez que llega un mensaje a la cola."""
    pedido = json.loads(body.decode("utf-8"))
    print(f"[ConsumidorExpress] Pedido recibido: {json.dumps(pedido)}")
    procesar_pedido_express(pedido)


def main():
    # 1. Conectar a RabbitMQ
    connection = pika.BlockingConnection(pika.ConnectionParameters("localhost"))
    channel = connection.channel()

    # 2. Declarar el exchange (debe coincidir con el productor)
    channel.exchange_declare(exchange=EXCHANGE_NAME, exchange_type="topic")

    # 3. Declarar la cola y vincularla al exchange con la routing key
    channel.queue_declare(queue=QUEUE_NAME)
    channel.queue_bind(exchange=EXCHANGE_NAME, queue=QUEUE_NAME, routing_key=ROUTING_KEY)

    # 4. Registrar el callback y empezar a consumir
    channel.basic_consume(queue=QUEUE_NAME, on_message_callback=callback, auto_ack=True)

    print("[ConsumidorExpress] Esperando pedidos express. CTRL+C para salir.")
    channel.start_consuming()


if __name__ == "__main__":
    main()
